### PR TITLE
Private operators - IO1-2147

### PR DESCRIPTION
**Pull request by mtabasco**
There should be an ability for operator to register privately

- Private operators are able to define a whitelist in which we store the owner address of validators
- getOperatorById will return private flag
- Operator functions to remove and update the whitelist address

### CODE REVIEW

File: contracts/SSVNetwork.sol [hunk 1]
```diff
@@ -241,10 +257,19 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
                     if (operator.snapshot.block == 0) {
                         revert OperatorDoesNotExist();
                     }
+                    if (
+                        operatorsWhitelist[operatorIds[i]] != address(0) &&
```
Comment (lior-blox) [inline contracts/SSVNetwork.sol:261]: @mtabasco lets check gas when the white list is in operator struct

File: contracts/SSVNetwork.sol [hunk 2]
```diff
@@ -797,6 +822,16 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
         emit OperatorRemoved(operatorId);
     }
 
+    function _removeOperatorWhitelist(uint64 operatorId, Operator storage operator) private onlyOperatorOwner(operator) {
```
Comment (lior-blox) [inline contracts/SSVNetwork.sol:825]: add to gas usage

File: contracts/SSVNetwork.sol [hunk 3]
```diff
@@ -797,6 +822,16 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
...
+        delete operatorsWhitelist[operatorId];
+        emit OperatorWhitelistRemoved(operatorId);
+    }
+
+    function _updateOperatorWhitelist(uint64 operatorId, address whitelisted, Operator storage operator) private onlyOperatorOwner(operator) {
```
Comment (lior-blox) [inline contracts/SSVNetwork.sol:830]: add to gas usage

**Comment by mtabasco** [comment]
Gas usage comparison
mapping ------struct property
REGISTER_VALIDATOR_NEW_STATE ---- 218300 218889
REGISTER_VALIDATOR_NEW_STATE_13  ----   429700 431760
@lior-blox

**Comment by mtabasco** [comment]
GAS USAGE

**mapping**
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD]: 194700
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE]: 211500
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT]: 174200
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD_7]: 265000
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_7]: 281800
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT_7]: 244500
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD_13]: 406000
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_13]: 422850
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT_13]: 385500

**struct**
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD]: 195400
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE]: 212300
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT]: 174900
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD_7]: 266200
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_7]: 283000
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT_7]: 245700
  [GasGroup.REGISTER_VALIDATOR_EXISTING_POD_13]: 408300
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_13]: 425100
  [GasGroup.REGISTER_VALIDATOR_NEW_STATE_WITHOUT_DEPOSIT_13]: 387800
  
  @lior-blox

File: contracts/SSVNetwork.sol [hunk 4]
```diff
@@ -127,7 +128,8 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
 
     function registerOperator(
         bytes calldata publicKey,
-        uint256 fee
+        uint256 fee,
+        address whitelisted
```
Comment (andrew-blox) [inline contracts/SSVNetwork.sol:132]: Doesnt this mean that we will have to deploy new contract?  `address whitelisted` will be new var or is it that you can upgrade and all existing ones will be null?

File: contracts/ISSVNetwork.sol [hunk 5]
```diff
@@ -167,7 +185,8 @@ interface ISSVNetwork is ISSVNetworkCore {
      */
     function registerOperator(
         bytes calldata publicKey,
-        uint256 fee
+        uint256 fee,
+        address whitelisted
```
Comment (andrew-blox) [inline contracts/ISSVNetwork.sol:189]: This means that all users will have to put 0x00000...?  They cant just put 0 right?  If so its a bit of overhead for the average user who wants to register a non private operator.

File: contracts/SSVNetworkViews.sol [hunk 6]
```diff
@@ -70,17 +70,17 @@ contract SSVNetworkViews is
...
+    ) external view override returns (address, uint256, uint32, bool, bool) {
         (
             address operatorOwner,
             uint64 fee,
             uint32 validatorCount,
             Snapshot memory snapshot
         ) = _ssvNetwork.operators(operatorId);
-        bool active;
-        if (snapshot.block != 0) active = true;
+        bool isPrivate = _ssvNetwork.operatorsWhitelist(operatorId) == address(0) ? false : true;
```
Comment (andrew-blox) [inline contracts/SSVNetworkViews.sol:80]: not sure this is enough.  What if an operator is private (has a fee of 0) then removes his address, we will say that he is not private even though he is

File: test/operators/register.ts [hunk 7]
```diff
@@ -18,56 +18,89 @@ describe('Register Operator Tests', () => {
...
-    )).to.emit(ssvNetworkContract, 'OperatorAdded').withArgs(1, helpers.DB.owners[1].address, publicKey, helpers.CONFIG.minimalOperatorFee);
+      ethers.constants.AddressZero
+    )).to.emit(ssvNetworkContract, 'OperatorAdded').withArgs(1, helpers.DB.owners[1].address, publicKey, helpers.CONFIG.minimalOperatorFee, ethers.constants.AddressZero);
+  });
+
+  it('Register private operator emits "OperatorAdded"', async () => {
+    const publicKey = helpers.DataGenerator.publicKey(0);
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[1]).registerOperator(
+      publicKey,
+      helpers.CONFIG.minimalOperatorFee,
```
Comment (andrew-blox) [inline test/operators/register.ts:29]: im confused :) i thought private operators must have fee of 0

**Review by andrew-blox — CHANGES_REQUESTED** [review]

File: contracts/SSVNetwork.sol [hunk 8]
```diff
@@ -145,7 +147,20 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
...
+        }
+
+        emit OperatorAdded(id, msg.sender, publicKey, fee, whitelisted);
+    }
+
+    function removeOperatorWhitelist(uint64 operatorId) external override  {
+        _removeOperatorWhitelist(operatorId, operators[operatorId]);   
+    }
+
+    function updateOperatorWhitelist(uint64 operatorId, address whitelisted) external override  {
```
[the remark this answers is not in the document]
Comment (mtabasco) [reply contracts/SSVNetwork.sol:162]: If only operators with zero fee can be private, then the logic should change, of course. @lior-blox please confirm, thanks!

File: contracts/ISSVNetwork.sol (hunk 5 above)
Comment (andrew-blox) [reply contracts/ISSVNetwork.sol:190]: Thats great i wanted to request that just to save gas for the normal register operators not needing to pay for the extra param / checks!

File: contracts/SSVNetwork.sol [hunk 9]
```diff
@@ -125,27 +126,34 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
     /* Operator External Functions */
     /*******************************/
 
+    function registerPrivateOperator(
```
Comment (andrew-blox) [inline contracts/SSVNetwork.sol:129]: @lior-blox I like the idea of second function what do you think?

**Review by andrew-blox — APPROVED** [review]

**Review by andrew-blox — APPROVED** [review]

File: contracts/SSVNetwork.sol (hunk 9 above)
Comment (mtabasco) [reply contracts/SSVNetwork.sol]: @andrew-blox We decided to go with `registerOperator -> setOperatorWhitelist` to enable a private operator. thanks!

File: contracts/SSVNetwork.sol [hunk 10]
```diff
@@ -672,12 +686,33 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
...
         emit OperatorRemoved(operatorId);
     }
 
+    function _removeOperatorWhitelist(
+        uint64 operatorId,
+        Operator storage operator
+    ) private onlyOperatorOwner(operator) {
+        delete operatorsWhitelist[operatorId];
+        emit OperatorWhitelistRemoved(operatorId);
+    }
```
Comment (drewwilliams11111) [inline contracts/SSVNetwork.sol:705]: Do we need a revert if its already empty?

**Review by drewwilliams11111 — APPROVED** [review]

File: contracts/SSVNetwork.sol (hunk 10 above)
Comment (mtabasco) [reply contracts/SSVNetwork.sol:705]: That would be useful mostly alerting the user. In terms of gas, if users use it without knowing if its set or not, then its more efficient to add the revert to not delete undefined value. but it users know the operator is whitelisted (via `Views.getOperatorById` `private` flag, then its more efficient to go an delete (current implementation).

File: contracts/SSVNetwork.sol [hunk 11]
```diff
@@ -789,11 +813,17 @@ contract SSVNetwork is UUPSUpgradeable, Ownable2StepUpgradeable, ISSVNetwork {
         emit OperatorRemoved(operatorId);
     }
 
-    function _declareOperatorFee(
-        uint64 operatorId,
-        Operator memory operator,
-        uint256 fee
-    ) private onlyOperatorOwner(operator) {
+    function _removeOperatorWhitelist(uint64 operatorId, Operator storage operator) private onlyOperatorOwner(operator) {
```
[the remark this answers is not in the document]
Comment (andrew-blox) [reply contracts/SSVNetwork.sol]: @mtabasco you did this right?

File: contracts/SSVNetwork.sol (hunk 10 above)
Comment (andrew-blox) [reply contracts/SSVNetwork.sol:705]: Resolving this and unresolved what we decided to go with

**Comment by lior-blox** [comment]
@mtabasco 
lets use just setOperatorwhitelist.
removing the remove function.
remove will be based on setting the address to 0

**Pull request merged**